server/src/main/java/org/opensearch/common/logging/DeprecatedMessage.java

```diff
@@ -37,15 +37,32 @@
 import org.opensearch.common.Strings;
 import org.opensearch.common.collect.MapBuilder;
 
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
 /**
  * A logger message used by {@link DeprecationLogger}.
  * Carries x-opaque-id field if provided in the headers. Will populate the x-opaque-id field in JSON logs.
  */
 public class DeprecatedMessage extends OpenSearchLogMessage {
     public static final String X_OPAQUE_ID_FIELD_NAME = "x-opaque-id";
+    private static Set<String> keys = ConcurrentHashMap.newKeySet();
+    private String keyWithXOpaqueId;
```

Comment:
Both of these can be final (see comment below about a change to the reset method).